yggdrasilffi/src/lib.rs

```diff
@@ -56,257 +63,241 @@ impl From<serde_json::Error> for FFIError {
     }
 }
 
-#[no_mangle]
-pub extern "C" fn engine_new() -> *mut c_void {
-    let state = EngineState::default();
-    Box::into_raw(Box::new(state)) as *mut c_void
+fn get_engine<'a>(engine_ptr: *mut c_void) -> Result<&'a mut EngineState, FFIError> {
+    if engine_ptr.is_null() {
+        Err(FFIError::NullError("Null engine pointer".into()))
+    } else {
+        Ok(unsafe { &mut *(engine_ptr as *mut EngineState) })
+    }
 }
 
-#[derive(Serialize, Deserialize, PartialEq, Eq)]
-enum ResponseCode {
-    Error = -2,
-    NotFound = -1,
-    Ok = 1,
+fn get_str<'a>(ptr: *const c_char) -> Result<&'a str, FFIError> {
+    if ptr.is_null() {
+        Err(FFIError::NullError("Null pointer".into()))
+    } else {
+        unsafe { CStr::from_ptr(ptr).to_str().map_err(FFIError::from) }
+    }
 }
 
-/// Frees the memory allocated for the engine
+fn get_json<T: DeserializeOwned>(json_ptr: *const c_char) -> Result<T, FFIError> {
```

Comment:
Same as above - this is 100% an unsafe function and callers should know that

Reply:
Thanks, same as https://github.com/Unleash/yggdrasil/pull/47#discussion_r1348066338